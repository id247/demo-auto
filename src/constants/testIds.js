export default {
  button: 'button',
  select: 'select',
  selectButton: 'select-button',
  input: 'input'
};
