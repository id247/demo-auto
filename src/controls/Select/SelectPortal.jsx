import React from 'react';
import ReactDOM from 'react-dom';

let dropdownRoot = document.getElementById('dropdown-root');
if (!dropdownRoot) {
  dropdownRoot = document.createElement('div');
  dropdownRoot.setAttribute('id', 'dropdown-root');
  document.body.appendChild(dropdownRoot);
}

class SelectPortal extends React.Component {
  el = document.createElement('div');

  componentDidMount() {
    dropdownRoot.appendChild(this.el);
  }

  componentWillUnmount() {
    dropdownRoot.removeChild(this.el);
  }

  render() {
    return ReactDOM.createPortal(this.props.children, this.el);
  }
}

export default SelectPortal;
