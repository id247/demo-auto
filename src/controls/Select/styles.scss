@import "styles/variables.scss";

.select {
  &__button {
    width: 100%;
    border-radius: $radius-s;
    border: 1px solid $color-grey;
    background: #fff;
    height: 32px;
    text-align: left;
    padding: 0 30px 0 $space-s;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    position: relative;
    font-size: $font-size-s;

    &:after {
      content: "";
      width: 0;
      height: 0;
      border-style: solid;
      border-width: 6px 5px 0 5px;
      border-color: $color-grey transparent transparent transparent;
      position: absolute;
      top: 50%;
      right: $space-m;
      margin-top: -3px;
    }

    &--is-opened:after {
      border-width: 0 5px 6px 5px;
      border-color: transparent transparent $color-grey transparent;
    }
  }
}

.select-dropdown {
  position: absolute;
  background: #fff;
  border: 1px solid $color-grey;
  margin: $space-s 0;
  border-radius: $radius-s;
  overflow: auto;
  max-height: 159px;

  &__item {
    padding: $space-m $space-s;
    font-size: $font-size-s;
    border: 0;
    display: block;
    width: 100%;
    text-align: left;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    cursor: pointer;

    &:not(:first-child) {
      border-top: 1px solid $color-grey;
    }

    &--is-selected#{&}--is-selectable,
    &:hover {
      background-color: $color-orange;
      color: #fff;
      border-color: $color-orange;
    }

    &:focus {
      outline: 1px dotted;
      outline-offset: -1px;
    }
  }
}
