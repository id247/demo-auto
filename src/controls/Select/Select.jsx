import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { selectOptionsPropType } from 'constants/propTypes';
import testIds from 'constants/testIds';
import SelectPortal from './SelectPortal';
import './styles.scss';

class Select extends React.Component {
  state = this.getState();

  select = React.createRef();
  dropdown = React.createRef();
  options = [];

  componentDidUpdate(_, prevState) {
    const { isOpen } = this.state;

    const isJustOpened = isOpen && !prevState.isOpen;
    const isJustCloseed = !isOpen && prevState.isOpen;

    if (isJustOpened) {
      this.updatePosition();
      this.options[0].focus();
      this.addListeners();
    }

    if (isJustCloseed) {
      this.removeListeners();
    }
  }

  componentWillUnmount() {
    this.removeListeners();
  }

  addListeners() {
    document.addEventListener('mousedown', this.handleClickOutside);
    window.addEventListener('resize', this.handleResize);
  }

  removeListeners() {
    document.removeEventListener('mousedown', this.handleClickOutside);
    window.removeEventListener('resize', this.handleResize);
  }

  handleClickOutside = e => {
    if (
      this.dropdown.current &&
      !this.dropdown.current.contains(e.target) &&
      !this.select.current.contains(e.target)
    ) {
      this.setState({
        isOpen: false
      });
    }
  };

  updatePosition() {
    const bodyRect = document.body.getBoundingClientRect();
    const selectRect = this.select.current.getBoundingClientRect();
    const left = selectRect.left - bodyRect.left;
    const selectTop = selectRect.top - bodyRect.top;
    const dropdownTop = selectTop + selectRect.height;

    const bottomLile = -bodyRect.top + window.innerHeight;

    this.dropdown.current.style.left = `${left}px`;
    this.dropdown.current.style.width = `${selectRect.width}px`;

    if (bottomLile - dropdownTop > 160) {
      this.dropdown.current.style.top = `${dropdownTop}px`;
      this.dropdown.current.style.bottom = ``;
    } else {
      const dropdownBottom = bodyRect.height - selectTop;
      this.dropdown.current.style.bottom = `${dropdownBottom}px`;
      this.dropdown.current.style.top = ``;
    }
  }

  handleResize = () => {
    this.setState({
      isOpen: false
    });
  };

  getState() {
    const { defaultValue, options } = this.props;
    let selected;

    if (defaultValue !== undefined) {
      selected = defaultValue;
    } else if (options.length > 0) {
      selected = options[0].value;
    }

    return {
      isOpen: false,
      selected
    };
  }

  handleClick = e => {
    this.setState(state => ({
      isOpen: !state.isOpen
    }));
  };

  handleSelect = value => e => {
    e.persist();
    const { onChange, name } = this.props;

    this.setState(
      {
        selected: value,
        isOpen: false
      },
      () => {
        if (typeof onChange === 'function') {
          onChange({
            ...e,
            target: {
              ...e.target,
              name,
              value
            }
          });
        }
      }
    );
  };

  render() {
    const { options, name, id, defautlValue, ...props } = this.props;
    const { isOpen, selected } = this.state;

    const selectedOption = options.find(o => o.value === selected) || {};

    return (
      <div
        ref={this.select}
        className="select"
        name={name}
        id={id || name}
        data-testid={testIds.select}
        {...props}
      >
        <button
          type="button"
          data-testid={testIds.selectButton}
          className={classNames('select__button', {
            'select__button--is-opened': isOpen
          })}
          onClick={this.handleClick}
        >
          {selectedOption.title}
        </button>
        <input type="hidden" name={name} value={selected} />
        {isOpen && options && options.length && (
          <SelectPortal>
            <div className="select-dropdown" ref={this.dropdown}>
              {options.map((o, i) => (
                <button
                  ref={node => (this.options[i] = node)}
                  type="button"
                  className={classNames('select-dropdown__item', {
                    'select-dropdown__item--is-selectable': !o.isEmpty,
                    'select-dropdown__item--is-selected': o.value === selected
                  })}
                  key={o.value}
                  onClick={this.handleSelect(o.value)}
                >
                  {o.title}
                </button>
              ))}
            </div>
          </SelectPortal>
        )}
      </div>
    );
  }
}

Select.propTypes = {
  options: selectOptionsPropType.isRequired,
  name: PropTypes.string.isRequired,
  defaultValue: PropTypes.string,
  id: PropTypes.string,
  onChange: PropTypes.func
};

export default Select;
