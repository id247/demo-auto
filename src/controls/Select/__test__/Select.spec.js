import 'jest-dom/extend-expect';
import React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  bindElementToQueries
} from 'react-testing-library';
import testIds from 'constants/testIds';
import Select from '../';

afterEach(cleanup);

describe('Select component', () => {
  const name = 'name';
  const id = 'id';
  const options = [
    {
      title: 'title',
      value: 'value'
    },
    {
      title: 'title2',
      value: 'value2'
    }
  ];
  const defautlValue = 'value';

  test('Renders with defaults', async () => {
    const { container, getByTestId } = render(
      <Select
        name={name}
        id={id}
        options={options}
        defautlValue={defautlValue}
      />
    );

    const select = getByTestId(testIds.select);
    expect(select).toHaveClass('select');
    expect(select).toHaveAttribute('name', name);
    expect(select).toHaveAttribute('id', id);

    const selectButton = container.querySelector('button');
    expect(selectButton).toHaveTextContent('title');

    expect(container.firstChild).toMatchSnapshot();
  });

  test('opens and closes dropdown', async () => {
    render(
      <Select
        name={name}
        id={id}
        options={options}
        defautlValue={defautlValue}
      />
    );
    const { getByTestId } = bindElementToQueries(document.body);
    const selectButton = getByTestId(testIds.selectButton);

    fireEvent.click(selectButton);
    let dropdown = document.body.querySelector('.select-dropdown');
    const firstOption = dropdown.firstChild;
    expect(dropdown).toBeVisible();
    expect(firstOption).toHaveFocus();

    fireEvent.click(selectButton);
    dropdown = document.body.querySelector('.select-dropdown');
    expect(dropdown).toBeFalsy();
  });

  test('changes selected item, fires onChange event and closes dropdown', async () => {
    const onChange = jest.fn();
    render(
      <Select
        name={name}
        id={id}
        options={options}
        onChange={onChange}
        defautlValue={defautlValue}
      />
    );
    const { getByTestId, getByText } = bindElementToQueries(document.body);
    const selectButton = getByTestId(testIds.selectButton);

    fireEvent.click(selectButton);
    let dropdown = document.body.querySelector('.select-dropdown');
    const secondOption = getByText('title2');

    fireEvent.click(secondOption);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(selectButton).toHaveTextContent('title2');

    dropdown = document.body.querySelector('.select-dropdown');
    expect(dropdown).toBeFalsy();
  });

  test('closes dropdown on click outside component', async () => {
    const { getByTestId } = render(
      <Select
        name={name}
        id={id}
        options={options}
        defautlValue={defautlValue}
      />
    );
    const selectButton = getByTestId(testIds.selectButton);

    fireEvent.click(selectButton);
    let dropdown = document.body.querySelector('.select-dropdown');

    fireEvent.mouseDown(document.body);

    dropdown = document.body.querySelector('.select-dropdown');
    expect(dropdown).toBeFalsy();
  });
});
