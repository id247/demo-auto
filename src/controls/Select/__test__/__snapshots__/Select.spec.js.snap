// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Renders with defaults 1`] = `
<div
  class="select"
  data-testid="select"
  id="id"
  name="name"
>
  <button
    class="select__button"
    data-testid="select-button"
    type="button"
  >
    title
  </button>
  <input
    name="name"
    type="hidden"
    value="value"
  />
</div>
`;
